README.md

[comment]: # (Start Badges)

[![Build Status](https://travis-ci.org/47deg/sbt-dependencies.svg?branch=master)](https://travis-ci.org/47deg/sbt-dependencies) [![Maven Central](https://img.shields.io/badge/maven%20central-0.4.0-green.svg)](https://repo1.maven.org/maven2/com/47deg/sbt-dependencies_2.12_1.0) [![License](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/47deg/sbt-dependencies/master/LICENSE) [![Join the chat at https://gitter.im/47deg/sbt-dependencies](https://badges.gitter.im/47deg/sbt-dependencies.svg)](https://gitter.im/47deg/sbt-dependencies?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![GitHub Issues](https://img.shields.io/github/issues/47deg/sbt-dependencies.svg)](https://github.com/47deg/sbt-dependencies/issues)

[comment]: # (End Badges)
# sbt-dependencies

**sbt-dependencies** is an SBT plugin that allows to you to keep your project dependencies up-to-date.

This plugin depends on the [`sbt-updates`](https://github.com/rtimush/sbt-updates) plugin

*Latest version*

[comment]: # (Start Replace)
```scala
addSbtPlugin("com.47deg" % "sbt-dependencies" % "0.4.0")
```

[comment]: # (End Replace)

## sbt-dependencies in the wild

If you wish to add your library here please consider a PR to include it.

[comment]: # (Start Copyright)
# Copyright

sbt-dependencies is designed and developed by 47 Degrees

Copyright (C) 2018 47 Degrees. <http://47deg.com>

[comment]: # (End Copyright)

CHANGELOG.md
# Changelog

## 12/14/2018 - Version 0.4.0

Release changes:

* Removes support for SBT 0.13.x ([#51](https://github.com/47deg/sbt-dependencies/pull/51))


## 11/16/2017 - Version 0.3.9

Release changes:

* Bumps sbt-org-policies ([#50](https://github.com/47deg/sbt-dependencies/pull/50))


## 11/16/2017 - Version 0.3.8

Release changes:

* Fixes sbt-org-policies-core version ([#47](https://github.com/47deg/sbt-dependencies/pull/47))
* Removes scala 2.10 from cross scala versions ([#49](https://github.com/47deg/sbt-dependencies/pull/49))


## 11/16/2017 - Version 0.3.7

Release changes:

* Fixes sbt-org-policies-core version ([#46](https://github.com/47deg/sbt-dependencies/pull/46))


## 11/15/2017 - Version 0.3.6

Release changes:

* Bumps sbt-org-policies, sbt-updates, and releases 0.3.6 ([#44](https://github.com/47deg/sbt-dependencies/pull/44))


## 10/23/2017 - Version 0.3.5

Release changes:

* Bumps sbt-org-policies version ([#40](https://github.com/47deg/sbt-dependencies/pull/40))
* Adds the org policies default settings to the projectSettings ([#39](https://github.com/47deg/sbt-dependencies/pull/39))
* Bumps sbt-org-policies plugin ([#41](https://github.com/47deg/sbt-dependencies/pull/41))
* Set crossScalaVersion to 2.12 and force org-policies-core version ([#42](https://github.com/47deg/sbt-dependencies/pull/42))
* Fixes GitHub token ([#43](https://github.com/47deg/sbt-dependencies/pull/43))


## 10/02/2017 - Version 0.3.4

Release changes:

* Enables microsite plugin again ([#38](https://github.com/47deg/sbt-dependencies/pull/38))


## 10/02/2017 - Version 0.3.3

Release changes:

* Makes site publication through Github4s ([#37](https://github.com/47deg/sbt-dependencies/pull/37))


## 10/02/2017 - Version 0.3.2

Release changes:

* Upgrade github4s sbt org policies ([#36](https://github.com/47deg/sbt-dependencies/pull/36))


## 09/19/2017 - Version 0.3.1

Release changes:

* Release 0.3.1 ([#35](https://github.com/47deg/sbt-dependencies/pull/35))


## 09/18/2017 - Version 0.3.0

Release changes:

* Release 0.3.0 ([#34](https://github.com/47deg/sbt-dependencies/pull/34))


## 09/13/2017 - Version 0.2.0

Release changes:

* Port plugin to sbt {1.0.x, 0.13.16} ([#30](https://github.com/47deg/sbt-dependencies/pull/30))
* Release 0.2.0 ([#31](https://github.com/47deg/sbt-dependencies/pull/31))
* Missing space in if condition ([#32](https://github.com/47deg/sbt-dependencies/pull/32))
* Remove depUpdateDependencyIssues in travis.yml ([#33](https://github.com/47deg/sbt-dependencies/pull/33))


## 04/25/2017 - Version 0.1.2

Release changes:

* Fixes publish microsites command ([#28](https://github.com/47deg/sbt-dependencies/pull/28))
* Upgrades sbt-org-policies ([#29](https://github.com/47deg/sbt-dependencies/pull/29))


## 04/20/2017 - Version 0.1.1

Release changes:

* Modifies the file committing ([#25](https://github.com/47deg/sbt-dependencies/pull/25))
* Removes pull command ([#26](https://github.com/47deg/sbt-dependencies/pull/26))
* Fixes git remote command ([#24](https://github.com/47deg/sbt-dependencies/pull/24))
* Adds a basic microsite ([#27](https://github.com/47deg/sbt-dependencies/pull/27))

src/main/scala/dependencies/Updates.scala
package dependencies

import com.timushev.sbt.updates.versions.Version
import sbt.ModuleID

import scala.collection.immutable.SortedSet

object Updates {

  def formatModule(module: ModuleID): String =
    module.organization + ":" + module.name + module.configurations.map(":" + _).getOrElse("")

  def patchUpdate(c: Version, updates: SortedSet[Version]): Option[Version] =
    updates.filter { v =>
      v.major == c.major && v.minor == c.minor
    }.lastOption

  def minorUpdate(c: Version, updates: SortedSet[Version]): Option[Version] =
    updates.filter { v =>
      v.major == c.major && v.minor > c.minor
    }.lastOption

  def majorUpdate(c: Version, updates: SortedSet[Version]): Option[Version] =
    updates.filter { v =>
      v.major > c.major
    }.lastOption

  def readUpdates(data: Map[ModuleID, SortedSet[Version]]): List[DependencyUpdate] =
    data.map {
      case (m, vs) =>
        val c = Version(m.revision)
        DependencyUpdate(
          moduleName = formatModule(m),
          revision = m.revision,
          patch = patchUpdate(c, vs).map(_.toString),
          minor = minorUpdate(c, vs).map(_.toString),
          major = majorUpdate(c, vs).map(_.toString)
        )
    }.toList.sortBy(_.moduleName)

}


src/main/scala/dependencies/GithubClient.scala
package dependencies

import cats.MonadError
import cats.instances.future
import github4s.Github
import github4s.Github._
import github4s.jvm.Implicits._
import github4s.free.domain._

import scala.concurrent.{ExecutionContext, Future}
import scalaj.http.HttpResponse

class GithubClient(owner: String, repo: String, accessToken: String)(implicit ec: ExecutionContext) {

  val issueTitle = "Update dependency"
  val issueLabel = "update-dependencies"

  implicit val monadError: MonadError[Future, Throwable] = catsStdInstancesForFuture

  def findIssuesByModuleName(): Future[Map[String, Issue]] = {

    def readIssues(issues: List[Issue]): Map[String, Issue] =
      issues.flatMap { issue =>
        if (issue.title.startsWith(issueTitle) && issue.title.length > issueTitle.length + 1) {
          val moduleName = issue.title.substring(issueTitle.length + 1)
          Option(moduleName -> issue)
        } else {
          None
        }
      }.toMap

    val searchParams = List(OwnerParamInRepository(s"$owner/$repo"),
                            IssueTypeIssue,
                            IssueStateOpen,
                            SearchIn(Set(SearchInTitle)),
                            LabelParam(issueLabel))

    Github(Some(accessToken)).issues
      .searchIssues(issueTitle, searchParams)
      .execFuture[HttpResponse[String]](Map("user-agent" -> "sbt-dependencies")) map {
      case Right(response) =>
        readIssues(response.result.items)
      case Left(e) => throw e
    }
  }

  def createIssue(dependencyUpdate: DependencyUpdate): Future[Issue] =
    Github(Some(accessToken)).issues
      .createIssue(owner = owner,
                   repo = repo,
                   title = title(dependencyUpdate),
                   body = body(dependencyUpdate),
                   labels = List(issueLabel),
                   assignees = List.empty)
      .execFuture[HttpResponse[String]](Map("user-agent" -> "sbt-dependencies")) map {
      case Right(response) => response.result
      case Left(e)         => throw e
    }

  def updateIssue(issue: Issue, dependencyUpdate: DependencyUpdate): Future[Issue] =
    Github(Some(accessToken)).issues
      .editIssue(owner = owner,
                 repo = repo,
                 issue = issue.number,
                 state = "open",
                 title = title(dependencyUpdate),
                 body = body(dependencyUpdate),
                 milestone = None,
                 labels = List(issueLabel),
                 assignees = issue.assignee.toList.map(_.login))
      .execFuture[HttpResponse[String]](Map("user-agent" -> "sbt-dependencies")) map {
      case Right(response) => response.result
      case Left(e)         => throw e
    }

  def title(dependencyUpdate: DependencyUpdate): String =
    s"$issueTitle ${dependencyUpdate.moduleName}"

  def body(dependencyUpdate: DependencyUpdate): String =
    s"""
       | Actual version: ${dependencyUpdate.revision}
       | Patch: ${dependencyUpdate.patch.getOrElse("-")}
       | Minor: ${dependencyUpdate.minor.getOrElse("-")}
       | Major: ${dependencyUpdate.major.getOrElse("-")}
         """.stripMargin

}

object GithubClient {

  def apply(owner: String, repo: String, accessToken: String)(
      implicit ec: ExecutionContext): GithubClient =
    new GithubClient(owner, repo, accessToken)

}


src/main/scala/dependencies/TablePrinter.scala
package dependencies

object TablePrinter {

  def format(table: List[List[String]]): String = table match {
    case Nil => ""
    case _ =>
      val sizes = table map {
        _.map { cell =>
          Option(cell) map (_.length) getOrElse 0
        }
      }
      val colSizes = sizes.transpose map (_.max)
      val rows = table.zipWithIndex.map {
        case (row, index) => formatRow(row, colSizes, index == 0)
      }
      formatRows(rowSeparator(colSizes), rows)
  }

  def formatRows(rowSeparator: String, rows: List[String]): String =
    (rowSeparator ::
      rows.head ::
        rowSeparator ::
          rows.tail :::
            rowSeparator ::
              List()).mkString("\n")

  def formatRow(row: List[String], colSizes: List[Int], header: Boolean): String = {
    val cells = row.zip(colSizes) map {
      case (item, size) if size == 0 => ""
      case (item, size)              => ("%" + size + "s").format(item)
    }
    cells.mkString("|", "|", "|")
  }

  def rowSeparator(colSizes: List[Int]): String = colSizes map { "-" * _ } mkString ("+", "+", "+")

}


src/main/scala/dependencies/DependenciesPlugin.scala
package dependencies

import com.timushev.sbt.updates.UpdatesPlugin.autoImport._
import com.timushev.sbt.updates.versions.Version
import github4s.free.domain.Issue
import sbt.Keys._
import sbt._

import scala.collection.immutable.SortedSet
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.util.{Failure, Success}

object DependenciesPlugin extends AutoPlugin {

  object autoImport extends DependenciesKeys

  import DependenciesPlugin.autoImport._

  def readUpdates(data: Map[ModuleID, SortedSet[Version]], log: Logger)(
      f: (List[DependencyUpdate]) => Unit): Unit =
    Updates.readUpdates(data) match {
      case Nil  => log.info("\nNo dependency updates found\n")
      case list => f(list)
    }

  def createIssueForDep(dep: DependencyUpdate,
                        issues: Map[String, Issue],
                        githubClient: GithubClient,
                        log: Logger): Future[Issue] = {
    log.info(s"Preparing issue for module `${dep.moduleName}`\n")
    issues.get(dep.moduleName) match {
      case Some(issue) =>
        log.info(s"Found existing open issue (#${issue.number}), updating it\n")
        githubClient.updateIssue(issue, dep) map { issue =>
          log.info(s"Issue updated at: ${issue.html_url}")
          issue
        }
      case None =>
        log.info("Existing issue not found, creating a new one\n")
        githubClient.createIssue(dep) map { issue =>
          log.info(s"Issue created at: ${issue.html_url}")
          issue
        }
    }
  }

  lazy val defaultSettings = Seq(
    showDependencyUpdates := {
      readUpdates(dependencyUpdatesData.value, streams.value.log) {
        list =>
          val fullTable = List("Module", "Revision", "Patch", "Minor", "Major") +:
              list.map(
                dep =>
                  List(dep.moduleName,
                       dep.revision,
                       dep.patch.getOrElse(""),
                       dep.minor.getOrElse(""),
                       dep.major.getOrElse("")))
          streams.value.log.info("\nFound some dependency updates:\n")
          streams.value.log.info(TablePrinter.format(fullTable))
          streams.value.log.info("Execute `updateDependencyIssues` to update your issues\n")
      }
    },
    updateDependencyIssues := {
      readUpdates(dependencyUpdatesData.value, streams.value.log) {
        list =>
          streams.value.log.info("Reading GitHub issues\n")
          githubToken.value match {
            case accessToken if accessToken.nonEmpty =>
              val githubClient = GithubClient(githubOwner.value, githubRepo.value, accessToken)
              val result = for {
                issues <- githubClient.findIssuesByModuleName()
                createdIssues <- Future.sequence(
                  list map (dep =>
                              createIssueForDep(dep, issues, githubClient, streams.value.log)))
              } yield createdIssues
              result onComplete {
                case Success(createdIssues) =>
                  streams.value.log.info("GitHub issues created or updated\n")
                case Failure(e) =>
                  streams.value.log.error(s"Error creating issues")
                  e.printStackTrace()
              }
            case _ =>
              streams.value.log.info(
                """
                  | Can't read the access token, please set the GitHub token with the SBT setting key 'githubToken'
                  |
                  | For ex:
                  |
                  |  // build.sbt
                  |  githubToken := sys.props.get("githubToken").getOrElse("")
                  |
                  |  // Command line
                  |  `sbt -DgithubToken=XXXXXX`)
                  |
                  | You need to create a token in this page with the `repo` scope:
                  |  * https://github.com/settings/tokens/new?scopes=repo&description=sbt-dependencies
                  |
                  | """.stripMargin)
          }

      }
    },
    dependencyUpdatesExclusions := moduleFilter(organization = "org.scala-lang"),
    githubToken := ""
  )

  override val projectSettings: Seq[Setting[_]] = defaultSettings

}


project/ProjectPlugin.scala
import com.typesafe.sbt.site.SitePlugin.autoImport._
import microsites.MicrositesPlugin.autoImport._
import sbt.Keys._
import sbt._
import sbtorgpolicies.OrgPoliciesPlugin
import sbtorgpolicies.OrgPoliciesPlugin.autoImport._
import sbtorgpolicies.model._
import sbtorgpolicies.templates.badges._

object ProjectPlugin extends AutoPlugin {

  override def trigger: PluginTrigger = allRequirements

  override def requires: Plugins = OrgPoliciesPlugin

  object autoImport {

    lazy val pluginSettings = Seq(
      sbtPlugin := true,
      scalaVersion := scalac.`2.12`,
      crossScalaVersions := Seq(scalac.`2.12`),
      crossSbtVersions := Seq(sbtV.`0.13`, sbtV.`1.0`),
      libraryDependencies ++= Seq(%%("github4s"), %%("org-policies-core", "0.8.12")),
      addSbtPlugin("com.timushev.sbt" % "sbt-updates" % "0.3.3")
    )

    lazy val micrositeSettings = Seq(
      micrositeName := "sbt-dependencies",
      micrositeDescription := "An SBT plugin that allows to you to keep your project dependencies up-to-date",
      micrositeBaseUrl := "sbt-dependencies",
      micrositeDocumentationUrl := "/sbt-dependencies/docs/",
      micrositeGithubOwner := "47deg",
      micrositeGithubRepo := "sbt-dependencies",
      micrositeHighlightTheme := "darcula",
      micrositePalette := Map(
        "brand-primary"   -> "#005b96",
        "brand-secondary" -> "#03396c",
        "brand-tertiary"  -> "#011f4b",
        "gray-dark"       -> "#48474C",
        "gray"            -> "#8D8C92",
        "gray-light"      -> "#E3E2E3",
        "gray-lighter"    -> "#F4F3F9",
        "white-color"     -> "#FFFFFF"
      ),
      includeFilter in makeSite := "*.html" | "*.css" | "*.png" | "*.jpg" | "*.gif" | "*.js" | "*.swf" | "*.md"
    )

  }

  override def projectSettings: Seq[Def.Setting[_]] =
    Seq(
      name := "sbt-dependencies",
      scalaOrganization := "org.scala-lang",
      startYear := Option(2017),
      orgMaintainersSetting := List(Dev("fedefernandez", Some("Fede Fernández"))),
      orgBadgeListSetting := List(
        TravisBadge.apply(_),
        MavenCentralBadge.apply(_),
        LicenseBadge.apply(_),
        GitHubIssuesBadge.apply(_)
      )
    ) ++ orgDefaultSettings ++ shellPromptSettings
}
